### DISCUSSION TITLE
Option to have Schema Registry as the source of truth (read-only)

### DISCUSSION THREAD

**Discussion by iMacTia** [flipp-oss/deimos#149] [Ideas]
The Schema Registry is not considered the source of truth.
Deimos currently not only reads but also writes against the schema registry, based on the files found locally.
This can be particularly dangerous in case someone with an outdated schema attempts to produce a message because it will effectively roll back the schema version in the schema registry to an older one.
We would like to limit deimos to only read from the schema registry and use that during the validation of a message (e.g. compare the auto-generated ruby class and warn when the application is producing using an outdated schema) without updating the schema registry in the process

**Comment by dorner** [comment]
Hey!

The schema registry can have multiple versions of the schema, and that's fine as long as the versions are compatible with each other. Publishing with a different version should *not* be dangerous if you're following the rules (and in fact, you can set up the schema registry to error out if you even try to publish an incompatible change, which I highly recommend). I'm a bit less inclined to support the workflow you suggest here because it takes the power out of the hands of the producing service, which should know what it needs to produce.

**Reply by iMacTia** [reply]
OK so that means that `deimos` doesn't compare against the latest version, but it also checks past versions?
That would be enough to avoid the "rollback" scenario I'm presenting here.

I actually had this issue with a topic schema, where I produced a message with the wrong schema and that ended up overriding the one defined in the schema registry. But I guess the setup you mentioned there would have saved me from that situation.

**Reply by dorner** [reply]
When producing, Deimos will use whatever schema is in the repo. When consuming, it will use whatever version was used to encode. Avro messages are encoded with the ID of the schema it was using, which represents both the subject and the version. When decoding, it will always fetch the exact ID it was encoded with.
